### PR TITLE
8333394: C2: assert(bol->is_Opaque4() || bol->is_OpaqueInitializedAssertionPredicate()) failed: Opaque node of non-null-check or of Initialized Assertion Predicate

### PR DESCRIPTION
**Pull request by chhagedorn**
[JDK-8330386](https://bugs.openjdk.org/browse/JDK-8330386) added some additional asserts to ensure that we are dealing with Template Assertion Predicates and/or non-null-checks which both use `Opaque4` nodes. I've also improved the following bailout from `!= 2` to `< 2` since the original comment suggested that if there is a test without 2 inputs, it must be a dead test (i.e. I assumed a `ConNode`):

https://github.com/openjdk/jdk/blob/91101f0d4fc8e06d0d74e06361db6ac87efeeb8e/src/hotspot/share/opto/loopTransform.cpp#L1204-L1206

But apparently, we could also have the following, also dead, `If` with a condition with 3 inputs (`505 Phi`) created by `split_thru_phi`:

image

I therefore revert the check back to `!= 2` and improved the comment.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333394](https://bugs.openjdk.org/browse/JDK-8333394): C2: assert(bol-&gt;is_Opaque4() || bol-&gt;is_OpaqueInitializedAssertionPredicate()) failed: Opaque node of non-null-check or of Initialized Assertion Predicate (**Bug** - P3)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to [5eb94738](https://git.openjdk.org/jdk/pull/19517/files/5eb9473803159837c17893ea7dd770c730ed11d7)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19517/head:pull/19517` \
`$ git checkout pull/19517`

Update a local copy of the PR: \
`$ git checkout pull/19517` \
`$ git pull https://git.openjdk.org/jdk.git pull/19517/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19517`

View PR using the GUI difftool: \
`$ git pr show -t 19517`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19517.diff">https://git.openjdk.org/jdk/pull/19517.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19517#issuecomment-2144607913)

### CODE REVIEW

File: test/hotspot/jtreg/compiler/predicates/assertion/TestIfWithPhiInput.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @bug 8333394
+ * @summary Test bailout of range check policy with an If with a Phi as condition.
+ * @run main/othervm -XX:CompileCommand=compileonly,*TestIfWithPhiInput*::* -Xcomp
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/predicates/assertion/TestIfWithPhiInput.java:29]:
```suggestion
 * @run main/othervm -XX:CompileCommand=compileonly,*TestIfWithPhiInput*::* -Xcomp -XX:-TieredCompilation
```

Just to make sure C2 compilation is triggered.

**Review by rwestrel — APPROVED** [review]
Looks good to me.

File: test/hotspot/jtreg/compiler/predicates/assertion/TestIfWithPhiInput.java (hunk 1 above)
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/predicates/assertion/TestIfWithPhiInput.java]: Good point, added.

**Review by TobiHartmann — APPROVED** [review]

**Pull request closed without merging**